### PR TITLE
Disable IPFS WebUI on Android

### PR DESCRIPTION
**Pull request by AlexeyBarabash**
Resolves https://github.com/brave/brave-browser/issues/29697, https://github.com/brave/brave-browser/issues/25103

This PR removes resources and some native code for ipfs-internals web ui page on Android. It allows to save only 162KB for now. This result may be improved with fixing of https://github.com/brave/brave-browser/issues/28060 .

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
1. Make fresh installation
2. Open ipfs://bafybeies3odi24wyk3e22rnautr57tiuk3b56nxrd53fxgtvr37abmz5j4/66.png link
3. Ensure you can still see this page
image
4. Press `Use a public gateway`
5. Ensure you can see the image
6. Navigate to about:chrome-urls
7. Ensure there is no brave://ipfs-internals

### CODE REVIEW

File: components/resources/ipfs_strings.grdp [hunk 1]
```diff
@@ -1,44 +1,49 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <grit-part>
   <if expr="enable_ipfs">
-    <message name="IDS_IPFS_CONNECTED_PEERS_TITLE" desc="Number of peers connected">Connected peers:</message>
-    <message name="IDS_IPFS_ADDRESSES_CONFIG_TITLE" desc="Title of addresses config section">Addresses</message>
-    <message name="IDS_IPFS_REPO_STATS_TITLE" desc="Title of repo stats section">Repo Stats</message>
-    <message name="IDS_IPFS_NODE_INFO_TITLE" desc="Title of node info section">Node info</message>
+    <if expr="enable_ipfs_internals_webui">
```
Comment (AlexeyBarabash) [inline components/resources/ipfs_strings.grdp:4]: Below are the strings from this section:
https://github.com/brave/brave-core/blob/master/browser/ui/webui/brave_webui_source.cc#L713

File: browser/ipfs/import/ipfs_import_controller.cc [hunk 2]
```diff
@@ -294,10 +294,12 @@ void IpfsImportController::OnWebPageImportCompleted(
 
 void IpfsImportController::OnImportCompleted(const ipfs::ImportedData& data) {
   auto link = CreateAndCopyShareableLink(data);
+#if BUILDFLAG(ENABLE_IPFS_INTERNALS_WEBUI)
```
Comment (AlexeyBarabash) [inline browser/ipfs/import/ipfs_import_controller.cc:297]: This may look strange, but on desktop this flag is true, and on Android whole `IpfsImportController` class seems to be unused, but compiled. It looks it will be fixed with https://github.com/brave/brave-browser/issues/28060. 

My attempt to make it failed - https://github.com/brave/brave-core/compare/android_no_ipfs_webui_BAK01?expand=1#diff-87dbeae647f32aa0670c977a84d99be9862a63bf9cc1f0a8b191da86f1d1a21fR11 - with such change I could not open ipfs:// pages at all.

**Comment by AlexeyBarabash** [comment]
Converted to draft as the unit test has failed on Android and error is related to the PR

**Comment by AlexeyBarabash** [comment]
CI fails with some tests on Windows-x86/Windows-x64, which are not related to the PR. I am going to open the PR again for the review.

**Comment by cypt4** [comment]
Is disabling enable_ipfs_local_node on android is much harder to implement than using separate build flag?

File: components/ipfs/buildflags/buildflags.gni [hunk 3]
```diff
@@ -1,7 +1,16 @@
...
 declare_args() {
   enable_ipfs = true
 }
 
 declare_args() {
   enable_ipfs_local_node = enable_ipfs && !is_ios
 }
+
+declare_args() {
+  enable_ipfs_internals_webui = enable_ipfs && !is_android
```
[the remark this answers is not in the document]
Comment (AlexeyBarabash) [reply components/ipfs/buildflags/buildflags.gni:15]: As I know, on iOS there is a limited support of internal pages.

`enable_ipfs_local_node` on iOS is already set to `false`, so no point to have `enable_ipfs_internals_webui` set to `true`, as it brave://ipfs-internals manages the local node.

[the remark this answers is not in the document]
Comment (AlexeyBarabash) [reply components/ipfs/buildflags/buildflags.gni]: Updated PR https://github.com/brave/brave-core/pull/17999/commits/9f83d71bc3a09bfaf8da2325199074a52f3cece0, @kylehickinson,  @Brandon-T could you take a look on this commit?

**Comment by AlexeyBarabash** [comment]
@cypt4 
> Is disabling enable_ipfs_local_node on android is much harder to implement than using separate build flag?

At this PR I tried to remove the obviously unused or useless data from APK.

I mentioned above that direct disabling `enable_ipfs_local_node` build flags makes IPFS unusable on Android, it seems that the `enable_ipfs_local_node` flag disables some required classes when using external gateway. Looks like this is know issue: https://github.com/brave/brave-browser/issues/28060.

I don't know which classes should be under `enable_ipfs_local_node` and which should not. I can't say is it much harder or not. 

In anyway, when https://github.com/brave/brave-browser/issues/28060 will be fixed, current PR will still be useful, it will allow to exclude ipfs-internals page resources and handlers depending on the `enable_ipfs_local_node` flag, just by renaming `ENABLE_IPFS_INTERNALS_WEBUI` to `ENABLE_IPFS_LOCAL_NODE` and removing `enable_ipfs_internals_webui` flag.

File: components/ipfs/buildflags/buildflags.gni (hunk 3 above)
[the remark this answers is not in the document]
Comment (kylehickinson) [reply components/ipfs/buildflags/buildflags.gni]: Lgtm! You may want to also consider asserting non-ios in the ipfs ui gn file like you do for android here but other then that all good!

[the remark this answers is not in the document]
Comment (AlexeyBarabash) [reply components/ipfs/buildflags/buildflags.gni]: Thanks! Added assert for non-iOS at https://github.com/brave/brave-core/pull/17999/commits/3ec749a4712fe2147370ea0d2fcac0ca0fe78581

File: components/ipfs_ui/BUILD.gn [hunk 4]
```diff
@@ -1,17 +1,23 @@
...
 
+assert(!is_android && !is_ios)
+
 transpile_web_ui("ipfs_ui") {
-  entry_points = [
-    ["ipfs", rebase_path("ipfs.tsx")]
-  ]
+  entry_points = [ [
+        "ipfs",
+        rebase_path("ipfs.tsx"),
```
Comment (cypt4) [inline components/ipfs_ui/BUILD.gn:13]: nit : some unnecessary formatting in this file

**Review by cypt4 — APPROVED** [review]
LGTM

File: chromium_src/chrome/common/webui_url_constants.cc [hunk 5]
```diff
@@ -4,17 +4,26 @@
...
 #define kChromeUIPerformanceSettingsURL kChromeUIPerformanceSettingsURL_UnUsed
 #define kPerformanceSubPage kPerformanceSubPage_UnUsed
 
+#if BUILDFLAG(ENABLE_IPFS_INTERNALS_WEBUI)
+#define kChromeUIBlobInternalsHost kChromeUIBlobInternalsHost, kIPFSWebUIHost
+#endif
+
 #include "src/chrome/common/webui_url_constants.cc"
 
+#if BUILDFLAG(ENABLE_IPFS_INTERNALS_WEBUI)
```
Comment (goodov) [inline chromium_src/chrome/common/webui_url_constants.cc:23]: I have recently found out that `#undef` is a no-op if a define doesn't exist, so we don't need to wrap `#undef` declarations into  `#if BUILDFLAG` and similar conditions.

File: components/ipfs_ui/BUILD.gn (hunk 4 above)
Comment (goodov) [reply components/ipfs_ui/BUILD.gn:13]: we use `gn format`, so some weird formatting is possible, but at least it is consistent.

**Review by goodov — APPROVED** [review]
`chromium_src`, `DEPS` lgtm

File: chromium_src/chrome/common/webui_url_constants.cc (hunk 5 above)
Comment (AlexeyBarabash) [reply chromium_src/chrome/common/webui_url_constants.cc]: Thanks, I didn't know this and I was sure it would give the error.

Fixed at https://github.com/brave/brave-core/pull/17999/commits/21aefb75ac7fc3e6ee47648d3cf20b70311b2a5d, it looks better now.

**Comment by AlexeyBarabash** [comment]
Squashed and rebased

**Review by mkarolin — APPROVED** [review]
strings ++

**Comment by AlexeyBarabash** [comment]
Rebased and resolved conflict

**Comment by AlexeyBarabash** [comment]
Rebased again.
